packages/node-rnw-rpc/src/nodeRnwRpc.ts

```diff
@@ -0,0 +1,156 @@
+/**
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT License.
+ *
+ * @format
+ */
+
+import {Server, Socket} from 'net';
+import * as jsonrpc from 'jsonrpc-lite';
+import {RpcStatusType} from 'jsonrpc-lite';
+
+let incrementingId = 0;
+
+export type InvokeResult =
+  | {type: 'error'; code: any; message: string}
+  | {type: 'success'; result?: any};
+
+export class RpcClient {
+  private readonly socket: Socket;
+  private readonly server: Server;
+  private readonly pendingRequests: Map<
+    any,
+    (result: InvokeResult | null, err: Error | null) => void
+  >;
+  private receiveBuffer: Buffer;
+
+  constructor(socket: Socket, server: Server) {
+    this.socket = socket;
+    this.server = server;
+
+    this.pendingRequests = new Map();
+    this.receiveBuffer = Buffer.alloc(0);
+
+    this.socket.on('data', this.onData.bind(this));
+    this.socket.on('end', this.onEnd.bind(this));
+    this.socket.on('error', this.onError.bind(this));
+  }
+
+  invoke(
+    methodName: string,
+    params: any[] | Record<string, any>,
+  ): Promise<InvokeResult> {
+    return new Promise((resolve, reject) => {
+      const messageId = ++incrementingId;
+      this.pendingRequests.set(messageId, (result, err) => {
+        if (err) {
+          reject(err);
+        }
+        if (result) {
+          resolve(result);
+        }
+      });
+
+      const requestString = jsonrpc
+        .request(messageId, methodName, params)
+        .serialize();
+
+      const requestBuffer = Buffer.from(requestString, 'utf-8');
+      const sizeBuffer = Buffer.alloc(4);
+      sizeBuffer.writeUInt32LE(requestBuffer.length);
+
+      this.socket.write(sizeBuffer);
+      this.socket.write(requestBuffer);
+    });
+  }
+
+  close() {
+    this.socket.destroy();
+    this.server.close();
+  }
+
+  private onData(chunk: Buffer) {
+    this.receiveBuffer = Buffer.concat([this.receiveBuffer, chunk]);
+    if (this.receiveBuffer.length >= 4) {
+      const messageLength = this.receiveBuffer.readUInt32LE();
+      const totalLength = messageLength + 4;
+
+      if (this.receiveBuffer.length >= totalLength) {
+        const messageBuffer = this.receiveBuffer.slice(4, totalLength);
+
+        if (totalLength < this.receiveBuffer.length) {
```

Comment:
> (totalL [](start = 10, length = 8)

i guess defense in depth, but would this be equal to a more readible `messageLength > 0`

Reply:
They measure different things here. `messageLength` is the length of the body, minus any headers, etc. (though `bodyLength` might be better). totalLength is the size of the buffer received, which may include partial/multiple messages.

Need to double check whether Node TCP onData is packet or buffer size. Docs aren't clear, but IIRC the latter. Code tries to be stream friendly, so allows same data chunk to stop and start.

Reply:
thanks, fully understood...